KIPs/kip-247.md

```diff
@@ -0,0 +1,186 @@
+    function getAmountIn(address token, uint amountOut) external view returns (uint amountIn);
+}
+```
+
+`swapForGas()` must ensure that all the following conditions are satisfied:
+
+- R1: Sender has enough tokens. That is, `token.balanceOf(msg.sender) >= amountIn`.
+- R2: Token is whitelisted and has a corresponding DEX address other than zero.
+- R3: `amountReceived >= minAmountOut >= amountRepay` where `amountReceived` is the swap output.
+
+`swapForGas()` must distribute the swap output (`amountReceived`) as follows:
+
+- The block proposer (`block.coinbase`) receives `amountRepay`.
+- User receives `(amountReceived - amountRepay) * (1 - commissionRate)`.
+- The commission `(amountReceived - amountRepay) * commissionRate` is stored in the contract for later claim.
+
+### TxPool Policy Updates
+
+#### GaslessTx Validation (validateTx)
+
+To prevent GaslessTx to be discarded, the validation logic at `txpool.queue` insertion needs to detect GaslessTx and skip the account balance check.
+
+```
+// txpool.validateTx()
+if ...
+else: // if tx is NOT feeDelegatedTx
+    if senderBalance.Cmp(tx.Cost()) < 0:
+        if tx is GaslessTx:
+            validation ok
+        else:
+            insufficientFund error
+```
+
+#### Promoting GaslessTx (promoteExecutables)
+
+GaslessApproveTx cannot be executed unless the corresponding GaslessSwapTx from the same sender is already in the transaction pool.
+Conversely, if the token has not been approved, GaslessSwapTx cannot be executed before GaslessApproveTx.
+In such cases, both transactions need to be promoted simultaneously.
+
+```
+// txpool.promoteExecutables()
+if tx is GaslessApproveTx:
+    if GaslessSwapTx is in queue:
+        promote GaslessApproveTx and GaslessSwapTx to pending
+    else:
+        push GaslessApproveTx to queue
+
+if tx is GaslessSwapTx:
+    if tx is GaslessApproveTx is in queue:
+        promote GaslessApproveTx and GaslessSwapTx to pending
+    else if GaslessSwapTx.nonce == getNonce(user):
+        push GaslessApproveTx to pending
+    else if GaslessSwapTx.nonce == getNonce(user) + 1:
+        promote GaslessSwapTx to queue
+```
+
+#### Transaction Bundling
+
+This KIP implements `TxBundlingModule` specified in [KIP-245](https://kips.kaia.io/KIPs/kip-245). These are the possible bundles:
+
+- LendTx - GaslessApproveTx - GaslessSwapTx
+- LendTx - GaslessSwapTx
+
+```
+ExtractTxBundles(txs, prevBundles):
+    if GaslessTx is disabled:
+        return nothing
+
+    filter ApproveTx, SwapTx from txs
+    move ApproveTx (if exists) before SwapTx
+    inject LendTx before ApproveTx if exists, otherwise before SwapTx
+```
+
+## Rationale
+
+### Disabling GaslessTx
+
+If many consensus nodes disabled GaslessTx feature, users may experience increased transaction latency.
+However, as long as there are still nodes supporting GaslessTx, the transactions will eventually be executed.
+Also, because this feature is enabled by default, few consensus nodes are expected to disable it.
+
```

Comment:
Please explain why state-dependent checks are not performed in the txpool. For instance, state-dependent checks cannot be accurate until the very last minute (i.e. just before executing the transaction) so we resort to bundling rollback even if we had checked them.

Reply:
added 0b0a68a